src/recommendationservice/Dockerfile

```diff
@@ -12,21 +12,33 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM python:3.10
+FROM python:3.10-slim-bullseye as base
+
+FROM base as builder
+
+RUN apt-get -qq update \
+    && apt-get install -y --no-install-recommends g++ \
+    && rm -rf /var/lib/apt/lists/*
 
 WORKDIR /usr/src/app/
 
 COPY ./src/recommendationservice/ ./
 COPY ./pb/ ./proto/
 
 RUN pip install --upgrade pip
-RUN pip install -r ./requirements.txt
-
-RUN opentelemetry-bootstrap -a install
+RUN pip install --prefix="/reqs" -r requirements.txt
 
 # add files into working directory
 RUN python -m pip install grpcio-tools==1.48.2
 RUN python -m grpc_tools.protoc -I=./proto/ --python_out=./ --grpc_python_out=./ ./proto/demo.proto
 
+FROM base
+
+WORKDIR /usr/src/app/
+COPY --from=builder /usr/src/app/ .
```

Comment:
You should be able to squeeze out some build time by doing the installation of the requirements and the grpc tools run in two separate stages that then can be run in parallel, e.g:

```Dockerfile
FROM base as grpc-builder

WORKDIR /usr/src/app/

COPY ./pb/ ./proto/

# add files into working directory
RUN python -m pip install grpcio-tools==1.48.2
RUN python -m grpc_tools.protoc -I=./proto/ --python_out=./ --grpc_python_out=./ ./proto/demo.proto

FROM base

WORKDIR /usr/src/app/
COPY --from=grpc-builder /usr/src/app/ .
COPY --from=builder /reqs /usr/local
```

Gives me ~10 seconds (from ~35 down to 25)

Reply:
Awesome!
Thanks for the tip @svrnm! 🤓